pkg/csr/csr.go

```diff
@@ -0,0 +1,355 @@
+	if a.isNodeClientCert(parsedCSR) {
+		// Node client bootstrapper CSR is received before the instance becomes a node
+		// hence we should not proceed if a corresponding node already exists
+		node := &core.Node{}
+		err := a.client.Get(context.TODO(), kubeTypes.NamespacedName{Namespace: a.namespace,
+			Name: nodeName}, node)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return false, errors.Wrapf(err, "unable to get node %s", nodeName)
+		} else if err == nil {
+			return false, errors.Wrapf(err, "%s node already exists, cannot validate CSR: %s", nodeName, a.csr.Name)
+		}
+	} else {
+		if err := a.validateKubeletServingCSR(parsedCSR); err != nil {
+			return false, errors.Wrapf(err, "unable to validate kubelet serving CSR: %s", a.csr.Name)
+		}
+	}
+	return true, nil
+}
+
+// validateNodeName returns true if the node name passed here matches either the
+// actual host name of the VM'S or the reverse lookup of the instance addresses
+// present in the configMap.
+func (a *Approver) validateNodeName(nodeName string) (bool, error) {
+	// Get the list of instances that are expected to be Nodes
+	windowsInstances, err := wiparser.GetInstances(a.client, a.namespace)
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to retrieve Windows instances")
+	}
+	// check if the node name matches the lookup of any of the instance addresses
+	hasEntry, err := matchesDNS(nodeName, windowsInstances)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to map node name to the addresses of Windows instances")
+	}
+	if hasEntry {
+		return true, nil
+	}
+	return a.validateWithHostName(nodeName, windowsInstances)
+}
+
+// validateWithHostName returns true if the node name given matches the host name for any of the instances
+// provided in the instance list. If a match is found, it also validates if the node name complies with the DNS
+// RFC1123 naming convention for internet hosts.
+func (a *Approver) validateWithHostName(nodeName string, windowsInstances []*instances.InstanceInfo) (bool, error) {
+	// Create a new signer using the private key secret
+	instanceSigner, err := signer.Create(kubeTypes.NamespacedName{Namespace: a.namespace,
+		Name: secrets.PrivateKeySecret}, a.client)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to create signer from private key secret")
+	}
+	// check if the node name matches any of the instances host names
+	hasEntry, err := matchesHostname(nodeName, windowsInstances, instanceSigner)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to map node name to the host names of Windows instances")
+	}
+	if !hasEntry {
+		// CSR is not from a BYOH instance
+		return false, nil
+	}
+	// validate node name for DNS RFC1123 naming conventions
+	// ref: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names
+	if errs := validation.IsDNS1123Subdomain(nodeName); len(errs) > 0 {
+		a.recorder.Eventf(a.csr, core.EventTypeWarning, "nodeNameValidationFailed",
```

Comment:
Sorry should have caught this earlier. `NodeNameValidationFailed` to be consistent with the existing event naming pattern.

Reply:
done